File: ngwidgets/leaflet.js
// https://github.com/zauberzeug/nicegui/blob/main/examples/map/leaflet.js
export default {
  template: "<div></div>",
  mounted() {
    this.map = L.map(this.$el);
    L.tileLayer("http://{s}.tile.osm.org/{z}/{x}/{y}.png", {
      attribution: '&copy; <a href="https://openstreetmap.org/copyright">OpenStreetMap contributors</a>',
    }).addTo(this.map);
    // Contour layer from OpenTopoMap
    L.tileLayer('https://{s}.tile.opentopomap.org/{z}/{x}/{y}.png', {
      maxZoom: 19,
      attribution: 'Map data: &copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap contributors</a>, <a href="https://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, Imagery © <a href="https://opentopomap.org">OpenTopoMap</a>',
      opacity: 0.7  // You can adjust the opacity to make the contour lines more or less pronounced
    }).addTo(this.map);
  },
  methods: {
    set_location(latitude, longitude,zoom_level) {
      this.target = L.latLng(latitude, longitude);
      this.map.setView(this.target, zoom_level);
      if (this.marker) {
        this.map.removeLayer(this.marker);
      }
      this.marker = L.marker(this.target);
      this.marker.addTo(this.map);
    },
    set_zoom_level(zoom_level) {
		this.map.setZoom(zoom_level);
	},
    draw_path(path) {
      if (this.pathLayer) {
        this.map.removeLayer(this.pathLayer);  // Remove previous path if any
      }
      this.pathLayer = L.polyline(path, {color: 'red'}).addTo(this.map);
      this.map.fitBounds(this.pathLayer.getBounds());  // Adjust view to fit the path
    }
  },
};

File: ngwidgets/leaflet.py
"""
(Re) Created on 2024-12-03

compatibility layer

@author: wf
"""
from typing import Dict, List, Tuple
from nicegui import ui

class leaflet(ui.leaflet):
    """Leaflet map wrapper for backwards compatibility with older nicegui versions"""

    def __init__(self, version: str = "1.7.1",zoom:int=9) -> None:
        """Initialize map with default center and zoom
        Args:
            version: Kept for backwards compatibility, not used"""
        super().__init__(center=(0.0, 0.0), zoom=zoom)
        self._path_layer = None

    def set_location(self, location: Tuple[float, float], zoom_level: int = None) -> None:
        """Set map center and zoom
        Args:
            location: (latitude, longitude) tuple
            zoom_level: Zoom level 1-18"""
        self.center = location
        if zoom_level:
            self.zoom = zoom_level

    def set_zoom_level(self, zoom_level: int) -> None:
        """Set map zoom level
        Args:
            zoom_level: Zoom level 1-18"""
        self.zoom = zoom_level

    def draw_path(self, path: List[Tuple[float, float]],options:Dict={'color': 'red', 'weight': 3, 'opacity': 0.7}) -> None:
        """Draw path on map
        Args:
            path: List of (lat, lon) points"""
        if hasattr(self, '_path_layer') and self._path_layer is not None:
            self.remove_layer(self._path_layer)
        path_config = {
            'latLngs': path,
            'options': options
        }
        self._path_layer = self.generic_layer(name='polyline', args=[path_config])